lib/absinthe/phase/document/execution/resolution.ex

```diff
@@ -285,30 +285,28 @@ defmodule Absinthe.Phase.Document.Execution.Resolution do
     |> propagate_null_trimming
   end
 
-  defp maybe_add_non_null_error([], nil, %Type.NonNull{}) do
+  defp maybe_add_non_null_error(errors, value, type, path \\ [])
+
+  defp maybe_add_non_null_error([], nil, %Type.NonNull{}, []) do
     ["Cannot return null for non-nullable field"]
   end
 
-  # Unwrap the non null so we can check again for the possible case of a non null
-  # inside of a list. We want that clause to handle both
-  # - `non_null(list_of(non_null(:thing)))`
-  # - `list_of(non_null(:thing))`
-  # Thus the single layer of unwrapping here.
-  defp maybe_add_non_null_error([], values, %Type.NonNull{of_type: type}) do
-    maybe_add_non_null_error([], values, type)
+  defp maybe_add_non_null_error([], nil, %Type.NonNull{}, path) do
+    [%{message: "Cannot return null for non-nullable field", path: path}]
+  end
+
+  defp maybe_add_non_null_error([], value, %Type.NonNull{of_type: %Type.List{} = type}, path) do
+    maybe_add_non_null_error([], value, type, path)
   end
 
-  defp maybe_add_non_null_error([], values, %Type.List{of_type: %Type.NonNull{}})
-       when is_list(values) do
-    values
-    |> Enum.with_index()
-    |> Enum.filter(&is_nil(elem(&1, 0)))
-    |> Enum.map(fn {_value, index} ->
-      %{message: "Cannot return null for non-nullable field", path: [index]}
+  defp maybe_add_non_null_error([], [_ | _] = values, %Type.List{of_type: type}, path) do
+    Enum.with_index(values)
+    |> Enum.flat_map(fn {value, index} ->
+      maybe_add_non_null_error([], value, type, path ++ [index])
```

Comment:
Appending to this list here will be `O(n^2)` and the lists could potentially be quite large. I'm working to refactor this.

Reply:
Oh very good catch, I breezed right past that!